lib/ukfs-virtiofs/virtiofs-fuse.c

```diff
@@ -0,0 +1,1057 @@
+		return ret;
+	if (ohdr) {
+		if (unlikely((size_t)ret < sizeof(*ohdr)))
+			return -EIO; /* Reply too short */
+		if (unlikely((size_t)ret != ohdr->len))
+			uk_pr_err("Response size mismatch: %zd != %u\n",
+				  ret, ohdr->len);
+		if (unlikely(ohdr->error))
+			return ohdr->error;
+	} else if (unlikely(ret)) {
+		return -EIO; /* Unexpected write */
+	}
+	return 0;
+}
+
+static inline
+size_t payload_len(ssize_t ret)
+{
+	UK_ASSERT(ret >= 0);
+	UK_ASSERT((size_t)ret >= sizeof(struct fuse_out_header));
+	return (size_t)ret - sizeof(struct fuse_out_header);
+}
+
+static inline
+int check_reply_payload(ssize_t ret, struct fuse_out_header *ohdr,
+			size_t expected)
+{
+	const int err = check_reply(ret, ohdr);
+
+	if (unlikely(err))
+		return err;
+	if (unlikely(payload_len(ret) != expected))
+		return -EIO;
+	return 0;
+}
+
+/* Lib-internal "public" API for FUSE ops */
+
+int virtiofs_fuse_destroy(struct virtiofs_fuse_dev *fdev)
+{
+	struct fuse_in_header hdr;
+	const struct uk_virtiofs_iovec iov[] = {
+		{ &hdr, sizeof(hdr) }
+	};
+	ssize_t rlen;
+
+	init_hdr(fdev, &hdr, FUSE_DESTROY, 0);
+	rlen = uk_virtiofs_request(fdev->dev, iov, 1, 0);
+	return check_reply(rlen, NULL);
+}
+
+#define VIRTIOFS_FUSE_FLAGS (FUSE_BIG_WRITES)
+
+int virtiofs_fuse_init(struct virtiofs_fuse_dev *fdev)
+{
+	struct fuse_in_header hdr;
+	struct fuse_init_in in = {
+		.major = FUSE_KERNEL_VERSION,
+		.minor = FUSE_KERNEL_MINOR_VERSION,
+		.flags = VIRTIOFS_FUSE_FLAGS,
+	};
+	struct fuse_out_header ohdr;
+	struct fuse_init_out out;
+	const struct uk_virtiofs_iovec iov[] = {
+		{ &hdr, sizeof(hdr) },
+		{ &in, sizeof(in) },
+		{ &ohdr, sizeof(ohdr) },
+		{ &out, sizeof(out) }
+	};
+	int err;
+	ssize_t rlen;
+	size_t paylen;
+	uint32_t minor;
+
+	init_hdr(fdev, &hdr, FUSE_INIT, sizeof(in));
+	rlen = uk_virtiofs_request(fdev->dev, iov, 2, 2);
+	if (unlikely((err = check_reply(rlen, &ohdr))))
+		return err;
+
+	paylen = payload_len(rlen);
+	/* Check major version */
+	if (unlikely(paylen < FIELDLEN(struct fuse_init_out, major)))
+		return -ENODEV; /* Reply w/o major number; bad device */
+	if (unlikely(out.major != FUSE_KERNEL_VERSION)) {
+		/* Device must not reply with a larger version than asked */
+		if (unlikely(out.major > FUSE_KERNEL_VERSION))
+			return -ENODEV;
+		/* We only support the current major version; this may change */
+		return -EPROTO;
+	}
+	/* Check minor version */
+	if (unlikely(paylen < FIELDLEN(struct fuse_init_out, minor)))
+		return -ENODEV; /* Correct major but no minor; bad device */
+	if (unlikely(out.minor < VIRTIOFS_MIN_FUSE_MINOR)) {
+		/* Init valid, but device is too old; destroy & error */
+		err = -EPROTO;
+		goto err_destroy;
+	}
+	minor = MIN(out.minor, (uint32_t)FUSE_KERNEL_MINOR_VERSION);
+
+	uk_pr_debug("Negociated FUSE version %u.%u\n",
+		    FUSE_KERNEL_VERSION, minor);
+
+	/* Check flags */
+	if (unlikely(paylen < FIELDLEN(struct fuse_init_out, flags)))
+		return -ENODEV;
+	if (unlikely(out.flags != VIRTIOFS_FUSE_FLAGS)) {
```

Comment:
I think the device is allowed to return more flags than requested (for backward compatibility). Maybe just `(out.flags & VIRTIOFS_FUSE_FLAGS) != VIRTIOFS_FUSE_FLAGS`?